<a href="https://colab.research.google.com/github/Tessellate-Imaging/monk_v1/blob/master/study_roadmaps/1_getting_started_roadmap/2_elemental_features_of_monk/5)%20Feature%20-%20Switch%20modes%20without%20reloading%20experiment%20-%20train%2C%20eval%2C%20infer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Goals

### 1. Understand how continuously and manually switching between training and externally validating help improve training

### 2. Experiment and learning how switching between train and val modes can help choose right hyper-parameters


### 2. Steps
    - You will use mxnet gluon backend for this example
    - You will first train a classifier using default params
    
    - You will switch mode from train to val for the first time here
    - You will then validate to check accuracy
    
    - You will switch mode from val to train here
    - You will reduce the learning rate (Need notfocus on how it is done for now)
    - You will retrain again using this new lr
    
    - You will switch mode from train to val for the second time here
    - You will then validate to check accuracy
    
    - You will again switch mode from val to train here
    - You will further change the learning rate (Need notfocus on how it is done for now)
    - You will retrain again using this newest lr
    
    - You will switch mode from train to val for the final time here
    - You will then validate to check accuracy

# Table of Contents


## [0. Install](#0)


## [1. Train a classifier using default settings](#1)


## [2. Switch mode from train to eval and validate](#2)


## [3. Switch back mode, reduce lr, retrain](#3)


## [4. Switch mode from train to eval and re-validate](#4)


## [5.  Switch back mode, change lr further, retrain](#5)


## [6. Switch mode from train to eval and re-validate](#6)

<a id='0'></a>
# Install Monk
  
 - git clone https://github.com/Tessellate-Imaging/monk_v1.git
 
 - cd monk_v1/installation/Linux && pip install -r requirements_cu9.txt
     - (Select the requirements file as per OS and CUDA version)

In [ ]:
!git clone https://github.com/Tessellate-Imaging/monk_v1.git

In [ ]:
# If using Colab install using the commands below
!cd monk_v1/installation/Misc && pip install -r requirements_colab.txt

# If using Kaggle uncomment the following command
#!cd monk_v1/installation/Misc && pip install -r requirements_kaggle.txt

# Select the requirements file as per OS and CUDA version when using a local system or cloud
#!cd monk_v1/installation/Linux && pip install -r requirements_cu9.txt

## Dataset - Malarial cell images
    - Credits: https://www.kaggle.com/iarunava/cell-images-for-detecting-malaria

In [ ]:
! wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$(wget --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=1mMEtGIK8UZNCrErXRJR-kutNTaN1zxjC' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=1mMEtGIK8UZNCrErXRJR-kutNTaN1zxjC" -O malaria_cell.zip && rm -rf /tmp/cookies.txt

In [ ]:
! unzip -qq malaria_cell.zip

In [ ]:
! wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$(wget --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=1WHpd7M-E_EiXmdjOr48BfvlUtMRPV6PM' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=1WHpd7M-E_EiXmdjOr48BfvlUtMRPV6PM" -O malaria_cell_val.zip && rm -rf /tmp/cookies.txt

In [2]:
! unzip -qq malaria_cell_val.zip

# Imports 

    - Using single mxnet-gluoncv backend for this tutorial

In [1]:
# Monk
import os
import sys
sys.path.append("monk_v1/monk/");

In [2]:
#Using mxnet-gluon backend 
from gluon_prototype import prototype

<a id='1'></a>
# Train a classifier using default settings

### Creating and managing experiments
        - Provide project name
        - Provide experiment name

In [3]:
gtf = prototype(verbose=1);
gtf.Prototype("Malaria-Cell", "exp-switch-modes");

Mxnet Version: 1.5.0

Experiment Details
    Project: Malaria-Cell
    Experiment: exp-switch-modes
    Dir: /home/abhi/Desktop/Work/tess_tool/gui/v0.3/finetune_models/Organization/development/v5.0_blocks/study_roadmap/change_post_num_layers/2_elemental_features_of_monk/workspace/Malaria-Cell/exp-switch-modes/



### This creates files and directories as per the following structure
    
    
    workspace
        |
        |--------Malaria-Cell
                        |
                        |
                        |-----exp-switch-modes
                                    |
                                    |-----experiment-state.json
                                    |
                                    |-----output
                                            |
                                            |------logs (All training logs and graphs saved here)
                                            |
                                            |------models (all trained models saved here)

### Load Dataset

In [4]:
gtf.Default(dataset_path="malaria_cell", 
            model_name="resnet18_v1", 
            num_epochs=5);

#Read the summary generated once you run this cell. 

Dataset Details
    Train path:     malaria_cell
    Val path:       None
    CSV train path: None
    CSV val path:   None

Dataset Params
    Input Size:   224
    Batch Size:   4
    Data Shuffle: True
    Processors:   4
    Train-val split:   0.7

Pre-Composed Train Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Pre-Composed Val Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num train images: 1411
    Num val images:   605
    Num classes:      2

Model Params
    Model name:           resnet18_v1
    Use Gpu:              True
    Use pretrained:       True
    Freeze base network:  True

Model Details
    Loading pretrained model
    Model Loaded on device
        Model name:                           resnet18_v1
        Num of potentially trainable layers:  41
        Num of actual trainable layer

monk_v1/monk/system/imports.py:193: UserWarning: ArgumentWarning: clipnorm and clipvalue are active only for keras in current version of Monk
  warnings.warn(msg)
monk_v1/monk/system/imports.py:193: UserWarning: ArgumentWarning: momentum_dampening_rate is active only for pytorch in current version of Monk
  warnings.warn(msg)


### From summary current Learning rate: 0.01

In [8]:
#Start Training
gtf.Train();

#Read the training summary generated once you run the cell and training is completed

Training Start
    Epoch 1/5
    ----------




    curr_lr - 0.01
    [Epoch 1] Train-acc: 0.709, Train-loss: 0.698 | Val-acc: 0.818182, Val-loss: 0.391, | time: 5.7 sec

    Epoch 2/5
    ----------




    curr_lr - 0.0098
    [Epoch 2] Train-acc: 0.753, Train-loss: 0.640 | Val-acc: 0.771901, Val-loss: 0.579, | time: 5.5 sec

    Epoch 3/5
    ----------




    curr_lr - 0.009604
    [Epoch 3] Train-acc: 0.772, Train-loss: 0.562 | Val-acc: 0.743802, Val-loss: 0.681, | time: 5.2 sec

    Epoch 4/5
    ----------




    curr_lr - 0.009411919999999999
    [Epoch 4] Train-acc: 0.788, Train-loss: 0.536 | Val-acc: 0.785124, Val-loss: 0.570, | time: 5.4 sec

    Epoch 5/5
    ----------




    curr_lr - 0.009223681599999999
    [Epoch 5] Train-acc: 0.814, Train-loss: 0.488 | Val-acc: 0.854545, Val-loss: 0.377, | time: 5.4 sec

    Training completed in: 0m 25s
    Best val Acc:          0.854545

Training End

Training Outputs
    Model Dir:   /home/abhi/Desktop/Work/tess_tool/gui/v0.3/finetune_models/Organization/development/v5.0_blocks/study_roadmap/elemental_features_of_monk/workspace/Malaria-Cell/exp-switch-modes/output/models/
    Log Dir:     /home/abhi/Desktop/Work/tess_tool/gui/v0.3/finetune_models/Organization/development/v5.0_blocks/study_roadmap/elemental_features_of_monk/workspace/Malaria-Cell/exp-switch-modes/output/logs/
    Final model: final
    Best model:  best_model
    Log 1 - Validation accuracy history log: val_acc_history.npy
    Log 2 - Validation loss history log:     val_loss_history.npy
    Log 3 - Training accuracy history log:   train_acc_history.npy
    Log 4 - Training loss history log:       train_loss_history.npy
    Log 5 - Training cu

<Figure size 432x288 with 0 Axes>

<a id='2'></a>
# Switch mode from train to eval and validate

In [9]:
gtf.Switch_Mode(eval_infer=True)

### Load the validation dataset

In [11]:
gtf.Dataset_Params(dataset_path="malaria_cell_val");
gtf.Dataset();

Dataset Details
    Test path:      malaria_cell_val
    CSV test path:  None

Dataset Params
    Input Size:  224
    Processors:   4

Pre-Composed Test Transforms
[{'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num test images: 1011
    Num classes:      2



### Run validation

In [12]:
accuracy, class_based_accuracy = gtf.Evaluate();

Testing



    Result
        class based accuracies
            0. Parasitized - 42.01183431952663 %
            1. Uninfected - 88.29365079365078 %
        total images:            1011
        num correct predictions: 658
        Average accuracy (%):    65.08407517309594



### Accuracy now is - 65.08% when learning rate is 0.01
(Can change when you run the exp)

<a id='3'></a>
# Switch back mode, reduce lr, retrain

In [16]:
gtf.Switch_Mode(train=True)

## Reduce learning rate from 0.01 to 0.001

In [17]:
# This part of code will be taken up again in upcoming sections
gtf.update_learning_rate(0.001);
gtf.Reload();

Update: Learning Rate - 0.001

Pre-Composed Train Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Pre-Composed Val Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num train images: 1411
    Num val images:   605
    Num classes:      2

Model Details
    Loading pretrained model
    Model Loaded on device
        Model name:                           resnet18_v1
        Num of potentially trainable layers:  61
        Num of actual trainable layers:       1



In [18]:
#Start Training
gtf.Train();

#Read the training summary generated once you run the cell and training is completed

Training Start
    Epoch 1/5
    ----------




    curr_lr - 0.009039207967999998
    [Epoch 1] Train-acc: 0.685, Train-loss: 0.610 | Val-acc: 0.765289, Val-loss: 0.506, | time: 5.6 sec

    Epoch 2/5
    ----------




    curr_lr - 0.008858423808639998
    [Epoch 2] Train-acc: 0.781, Train-loss: 0.471 | Val-acc: 0.796694, Val-loss: 0.452, | time: 5.8 sec

    Epoch 3/5
    ----------




    curr_lr - 0.008681255332467197
    [Epoch 3] Train-acc: 0.791, Train-loss: 0.448 | Val-acc: 0.826446, Val-loss: 0.404, | time: 5.8 sec

    Epoch 4/5
    ----------




    curr_lr - 0.008507630225817853
    [Epoch 4] Train-acc: 0.819, Train-loss: 0.412 | Val-acc: 0.828099, Val-loss: 0.394, | time: 6.2 sec

    Epoch 5/5
    ----------




    curr_lr - 0.008337477621301496
    [Epoch 5] Train-acc: 0.816, Train-loss: 0.417 | Val-acc: 0.804959, Val-loss: 0.414, | time: 5.5 sec

    Training completed in: 0m 51s
    Best val Acc:          0.828099

Training End

Training Outputs
    Model Dir:   /home/abhi/Desktop/Work/tess_tool/gui/v0.3/finetune_models/Organization/development/v5.0_blocks/study_roadmap/elemental_features_of_monk/workspace/Malaria-Cell/exp-switch-modes/output/models/
    Log Dir:     /home/abhi/Desktop/Work/tess_tool/gui/v0.3/finetune_models/Organization/development/v5.0_blocks/study_roadmap/elemental_features_of_monk/workspace/Malaria-Cell/exp-switch-modes/output/logs/
    Final model: final
    Best model:  best_model
    Log 1 - Validation accuracy history log: val_acc_history.npy
    Log 2 - Validation loss history log:     val_loss_history.npy
    Log 3 - Training accuracy history log:   train_acc_history.npy
    Log 4 - Training loss history log:       train_loss_history.npy
    Log 5 - Training cu

<Figure size 432x288 with 0 Axes>

<a id='4'></a>
# Switch mode from train to eval and re-validate

In [19]:
gtf.Switch_Mode(eval_infer=True)

### Load the validation dataset

In [21]:
gtf.Dataset_Params(dataset_path="malaria_cell_val");
gtf.Dataset();

Dataset Details
    Test path:      malaria_cell_val
    CSV test path:  None

Dataset Params
    Input Size:  224
    Processors:   4

Pre-Composed Test Transforms
[{'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num test images: 1011
    Num classes:      2



### Run validation

In [22]:
accuracy, class_based_accuracy = gtf.Evaluate();

Testing



    Result
        class based accuracies
            0. Parasitized - 28.79684418145957 %
            1. Uninfected - 89.0873015873016 %
        total images:            1011
        num correct predictions: 595
        Average accuracy (%):    58.852621167161224



### Accuracy now is - 58.85% when learning rate is 0.001 
(Can change when you run the exp)
    - Thus reducing learning rate didn't help our case

<a id='5'></a>
# Switch back mode, change lr, retrain

In [23]:
gtf.Switch_Mode(train=True)

## Update the learning rate again

In [24]:
# This part of code will be taken up again in upcoming sections
gtf.update_learning_rate(0.1);
gtf.Reload();

Update: Learning Rate - 0.1

Pre-Composed Train Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Pre-Composed Val Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num train images: 1411
    Num val images:   605
    Num classes:      2

Model Details
    Loading pretrained model
    Model Loaded on device
        Model name:                           resnet18_v1
        Num of potentially trainable layers:  61
        Num of actual trainable layers:       1



In [25]:
#Start Training
gtf.Train();

#Read the training summary generated once you run the cell and training is completed

Training Start
    Epoch 1/5
    ----------




    curr_lr - 0.008170728068875466
    [Epoch 1] Train-acc: 0.534, Train-loss: 0.803 | Val-acc: 0.659504, Val-loss: 0.606, | time: 5.9 sec

    Epoch 2/5
    ----------




    curr_lr - 0.008007313507497957
    [Epoch 2] Train-acc: 0.667, Train-loss: 0.612 | Val-acc: 0.761983, Val-loss: 0.534, | time: 5.6 sec

    Epoch 3/5
    ----------




    curr_lr - 0.007847167237347998
    [Epoch 3] Train-acc: 0.712, Train-loss: 0.559 | Val-acc: 0.761983, Val-loss: 0.481, | time: 5.5 sec

    Epoch 4/5
    ----------




    curr_lr - 0.007690223892601038
    [Epoch 4] Train-acc: 0.738, Train-loss: 0.517 | Val-acc: 0.745455, Val-loss: 0.498, | time: 5.3 sec

    Epoch 5/5
    ----------




    curr_lr - 0.007536419414749017
    [Epoch 5] Train-acc: 0.764, Train-loss: 0.490 | Val-acc: 0.808264, Val-loss: 0.426, | time: 5.4 sec

    Training completed in: 1m 17s
    Best val Acc:          0.808264

Training End

Training Outputs
    Model Dir:   /home/abhi/Desktop/Work/tess_tool/gui/v0.3/finetune_models/Organization/development/v5.0_blocks/study_roadmap/elemental_features_of_monk/workspace/Malaria-Cell/exp-switch-modes/output/models/
    Log Dir:     /home/abhi/Desktop/Work/tess_tool/gui/v0.3/finetune_models/Organization/development/v5.0_blocks/study_roadmap/elemental_features_of_monk/workspace/Malaria-Cell/exp-switch-modes/output/logs/
    Final model: final
    Best model:  best_model
    Log 1 - Validation accuracy history log: val_acc_history.npy
    Log 2 - Validation loss history log:     val_loss_history.npy
    Log 3 - Training accuracy history log:   train_acc_history.npy
    Log 4 - Training loss history log:       train_loss_history.npy
    Log 5 - Training cu

<Figure size 432x288 with 0 Axes>

<a id='6'></a>
# Switch mode from train to eval and re-validate

In [26]:
gtf.Switch_Mode(eval_infer=True)

### Load the validation dataset

In [27]:
gtf.Dataset_Params(dataset_path="malaria_cell_val");
gtf.Dataset();

Dataset Details
    Test path:      malaria_cell_val
    CSV test path:  None

Dataset Params
    Input Size:  224
    Processors:   4

Pre-Composed Test Transforms
[{'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num test images: 1011
    Num classes:      2



### Run validation

In [28]:
accuracy, class_based_accuracy = gtf.Evaluate();

Testing



    Result
        class based accuracies
            0. Parasitized - 53.84615384615385 %
            1. Uninfected - 45.83333333333333 %
        total images:            1011
        num correct predictions: 504
        Average accuracy (%):    49.85163204747774



### Accuracy now is - 49.85% when learning rate is 0.1, even  lower
(Can change when you run the exp)
    - Thus increasing learning rate didn't help our case

### LR 0.01 worked best for us
    - That's how manual hyper-parameter tuning can be done using switch modes